Kitodo/src/main/java/org/kitodo/production/helper/IndexWorker.java

```diff
@@ -50,91 +42,64 @@ public IndexWorker(SearchService searchService) {
-                if (amountToIndex > indexLimit) {
-                    amountToIndex = indexLimit;
-                }
-                while (this.indexedObjects < amountToIndex) {
-                    indexChunks(batchSize);
+        int timeBetweenAttempts = ConfigCore.getIntParameterOrDefaultValue(ParameterCore.ELASTICSEARCH_TIME_BETWEEN_ATTEMPTS);
+        int maxBatch = indexWorkerStatus.getMaxBatch();
+
+        Integer nextBatch = indexWorkerStatus.getAndIncrementNextBatch();
+        while (nextBatch < maxBatch) {
+            // nextBatch is a valid batch that needs to be processed
+
+            int attempt = 1;
+            while (attempt < maxAttempts) {
+                try {
+                    int offset = nextBatch * batchSize;
+                    logger.info("index " + objectType.toString() + " with offset " + offset + " and attempt " 
+                        + attempt + "/" + maxAttempts);
+
+                    if (indexAllObjects) {
+                        indexObjects(searchService.getAll(offset, batchSize));
+                    } else {
+                        indexObjects(searchService.getAllNotIndexed(offset, batchSize));
+                    }
+
+                    break;
+                } catch (Exception e) {
+                    logger.error(e.getMessage(), e);
+                    attempt += 1;
+                    try {
+                        Thread.sleep(timeBetweenAttempts);
+                    } catch (InterruptedException e2) {
+                        // ignore
+                    }
                 }
             }
-        } catch (CustomResponseException | DAOException  | HibernateException | IOException e) {
-            logger.error(e.getMessage(), e);
-        }
-    }
-
-    private int getAmountToIndex() throws DAOException {
-        if (indexAllObjects) {
-            return searchService.countDatabaseRows().intValue() - this.startIndexing;
-        } else {
-            return searchService.countNotIndexedDatabaseRows().intValue() - this.startIndexing;
-        }
-    }
 
-    @SuppressWarnings("unchecked")
-    private void indexChunks(int batchSize) throws CustomResponseException, DAOException, IOException {
-        List<Object> objectsToIndex;
-        int indexLimit = ConfigCore.getIntParameterOrDefaultValue(ParameterCore.ELASTICSEARCH_INDEXLIMIT);
-        while (this.indexedObjects < indexLimit) {
-            int offset = this.indexedObjects + this.startIndexing;
-
-            if (indexAllObjects) {
-                objectsToIndex = searchService.getAll(offset, batchSize);
-            } else {
-                objectsToIndex = searchService.getAllNotIndexed(offset, batchSize);
-            }
-            if (objectsToIndex.isEmpty()) {
-                break;
+            if (attempt >= maxAttempts) {
+                logger.error("stop indexing after maximum amount of attempts");
+                throw new RuntimeException("stop indexing after maximum amount of attempts");
```

Comment:
Maybe use a string to avoid duplicated text message? I don't know if logging and throwing an exception should happen on one place and `RuntimeException` must be catched explicit. Maybe @solth should decide here what is a good solution.

Reply:
The RuntimeException was used to break out of the while-loop. I added a boolean "failed" to gracefully exit the while loop and thread.